contracts/middle-layer/resource-mirror/MultiMessage.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../../interface/ICrossChain.sol";
+import "../../interface/IMultiMessage.sol";
+import "./CmnHub.sol";
+import "./storage/MultiStorage.sol";
+
+contract MultiMessage is MultiStorage, CmnHub, IMultiMessage {
+    uint8 public constant ACK_PACKAGE = 0x01;
+    uint8 public constant FAIL_ACK_PACKAGE = 0x02;
+
+    mapping(address => bool) public whitelistTargets;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    /*----------------- initializer -----------------*/
+    function initialize(address _ERC721_token, address _additional) public initializer {
+        __cmn_hub_init_unchained(_ERC721_token, _additional);
+
+        channelId = MULTI_MESSAGE_CHANNEL_ID;
+
+        whitelistTargets[BUCKET_HUB] = true;
+        whitelistTargets[GROUP_HUB] = true;
+        whitelistTargets[OBJECT_HUB] = true;
+        whitelistTargets[TOKEN_HUB] = true;
+    }
+
+    function initializeV2() public reinitializer(2) {
+        __cmn_hub_init_unchained_v2(INIT_MAX_CALLBACK_DATA_LENGTH);
+    }
+
+    function sendMessages(
+        address[] calldata _targets,
+        bytes[] calldata _data,
+        uint256[] calldata _values
+    ) external payable returns (bool) {
+        require(_targets.length == _data.length, "length mismatch");
+        require(_targets.length == _values.length, "length mismatch");
+
+        // generate packages
+        uint256 _totalValue = 0;
+        uint256 _totalRelayFee = 0;
+        uint256 _totalAckRelayFee = 0;
```

Comment:
please ensure the sum of `_values` is less than msg.value

Reply:
ensured `_totalValue == msg.value` in line 71